### DISCUSSION TITLE
Travis CI replacement for SWIG testing

### DISCUSSION THREAD

**Discussion by wsfulton** [swig/swig#1947] [General]
@vadz mentioned we might be able to migrate to GitHub Actions from Travis CI. I'd like to know some more details, even if it is just a good reference to Github documentation. One overriding concern that I have is this: can we get the equivalent to 100+ builds running like we do now on 3 different operating systems? That's quite a powerful build farm that we need to replace.

At the top of our builds on Travis, I now see "Please be aware travis-ci.org will be shutting down in several weeks, with all accounts migrating to travis-ci.com." Perhaps we should migrate to travis-ci.com?

**Comment by vadz** [comment]
Travis CI now heavily throttles open source projects and I think it must be even worse for SWIG which has a rather long running test suite. I don't think switching to `travis-ci.com` would change this, AFAIU only paying for a plan would.

IME GitHub Actions are _much_ faster than Travis, although, of course, nobody knows it it's going to last. The only drawback of it compared to Travis CI I've seen so far is that you can only run the builds there on a few system versions (a couple, or even just the, latest one(s) for each of Windows, Ubuntu and macOS). But this is probably not that important for SWIG and I guess we could run it inside Docker there if we really wanted to test under older systems too.

Reference documentation is [here](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions), and it might be enough as it's rather similar to Travis and AppVeyor that you already know, but if not, there is also an overview reachable from this link.

**Comment by wsfulton** [comment]
Unless you know of some other open source plan, https://github.com/features/actions indicates 2000 free minutes per month, but we use about 1000 minutes per commit. Running tests twice a months isn't going to cut it. This is my main concern.

**Reply by vadz** [reply]
2000 minutes limit is for private repositories on the Free plan only. It's "Free for public repositories", see https://github.com/pricing

Also, I haven't used it yet, but GitHub supports hosted runners, so it's always possible to run them on volunteers' machines while still using GitHub infrastructure, web UI and all that.

P.S. In any case I definitely don't know of any better alternative right now, if you do, please let me know.

**Comment by wsfulton** [comment]
Okay, that wording is confusing, well to me anyway! I'll assume it means unlimited free minutes then.

Travis runs 5 jobs concurrently. It isn't clear what Github will do. What do you see in your other orgs?

I'm all for trying out some new jobs using Github actions then. I suggest an updated MinGW run as these are currently broken on Appveyor and the Appveyor.yml is rather complicated trying to handle 3 similar OSs: Windows, Cygwin, MinGW. If you need me to set up some permissions/accounts, let me know, but we already have a Github org.

BTW, I'm working on getting the Travis OSX problems fixed.

**Reply by vadz** [reply]
It runs all my jobs concurrently, but I don't have as many as SWIG does in neither of the orgs I use GitHub Actions with, so maybe there are still some limits -- but definitely higher than 5, anyhow.

**Comment by wsfulton** [comment]
Sounds good. It would be quite a job switching over. The nice thing though is nothing stopping us from using both either long term or as a short term strategy. I'd be happy to support someone else to get this going.

BTW, I have managed to get Travis 100% passing again by updating to the latest MacOS images on Travis. There is still this migration from travis-ci.org to travis-ci.com which needs doing so I'll try have a look at that.

**Comment by rwf1** [comment]
As of 16 June 2021 Travis-ci.org is no longer doing builds.  So at this point the only testing that is going on is what is supported by appveyor.  This blog entry seems relevant to deciding how to go forward  [https://blog.travis-ci.com/2021-06-16-ibm-aws](https://blog.travis-ci.com/2021-06-16-ibm-aws)

**Comment by ojwb** [comment]
We're using GHA now, closing
